### PR TITLE
[3pt] PR: Add linting tools and updating files

### PR DESCRIPTION
**Pull request by RobHanna-NOAA**
We have upgraded all files to add a linting system using packages of `isort`, `black`, and `flake8 (pflake8)`.  All files were run through those three tools in that order, each one at a time with minor cleanup being done along the way.

Some functions existed in multiple files and some of them were moved to shared_functions for commonality / re-usability. 

Other minor changes were  most triggered by packages updates.

A new tool was also added called `hash_compare.py` which can take two files or two directories and using hashing to look for differences. This was used to validate changes via the addition of linting.

**Note**: A previously detected warning saying `PROJ: proj_create_from_database: Cannot find proj.db` continues to show up quite a bit. It does not hurt logic or code. A issue card already exists for it. [# 145](https://github.com/NOAA-OWP/ras2fim/issues/145).

In order to add the three new packages, it forced other package upgrades and some downgrades which triggered even more changes, so a full enviro reload is required. Here are the steps to reload the enviro.

### Conda Environment Upgrade
**Note** There is a critical upgrade to the ras2fim conda and it is not backwards compatible. To upgrade this particular version, 

1) We need to full uninstall the ras2fim environment, not upgraded it:
   - run: _conda deactivate_  (if you are already activated in ras2fim)
   - run:  _conda remove --name ras2fim --all_
   - Make sure you have downloaded (or merged) this PR (or dev branch).  Ensure you are in that folder in the `ras2fim` folder where the environment.yml is at.
   - run: _conda env create -f environment.yml_  . It might be slow, but 5 to 10 mins is reasonable.

### Code Merge Compatibility
There are literally hundreds of changes and it will be extremely hard to merge older code bases into this. It is strongly recommended to restart your changes by loading this new full branch, then manually add your changes back in to it.  Then run the three linting tools again before checking it it. Details lower.

### Linting Tools Usage
It is now expected that when you change a file, before you check it in, run the three linting tools using the following pattern. 
1) in cmd, run `isort {your file name}`.  e.g.  isort ras2fim.py.
2) in cmd, run `black {your file name}`.  e.g. black create_fim_rasters.py
3) Review the changes from black and make minor changes if you see fit. There are a good handful of anomalies and things that black will do that are not necessarily intuitive. Some of the major ones will be listed lower.
4) in cmd, run `pflake8 {your file name}`. e.g. pflake8 clip_dem_from_shape.py

We have added a configuration file that helps manage the rules we want in place for the three tools. The file is named pyproject.toml and is automatically used when using the three linting tools.  There are some tidbits, configuration and byproducts listed below.

### Linting Tidbits, Configuration and By-Products
- we set out line length at 110
- black does not attempt to line split when comments are in a line. You will need to split these yourselves to get it under 110. Note: If you have a line with some code and a comment after it, it will not split it correctly. Maybe go above and add extra blank line above that comment you like.  e.g.
    - was:
    ```
         some_var = "some string"
         print("do something but print cmd itself not over 100 …..") # and pretend this puts is over 110
    ```
    - now: consider (but optional as long as you don't break 110 in total)
    ```
        some_var = "some string"
		
	# and pretend this used to put it over 110 chars
        print("do something but print cmd itself not over 100 …..")
    ```
- Most of the time, argument values into a function will leave an extra comma on the end of the list and that is ok. e.g.
    ```
	some_var = ["some value",
                            "another value",]
    ```
- You may not be familiar with PEP-8 standards for `import`s and `from`s, but it is correct. See [isort](https://pycqa.github.io/isort/). 

### Additions  
- `pyproject.toml`: helps manage ras2fim rules for linting and it auto applied when using the linting commands in the same directory as current ras2fim code (your branch).
- `tools\hash_compare.py`: a tool to help compare two files or two directories using hashing to look for differences.

### Changes  
Almost all py files were changed for linting as well as some other additions listed above.

### Testing

Various tests for ras2fim.py including combinations of true/false for the new r2f_config.env. get_ras_models_by_catalog.py was also retested.

### Checklist

_You may update this checklist before and/or after creating the PR. If you're unsure about any of them, please ask, we're here to help! These items are what we are going to look for before merging your code._

- [x] Informative and human-readable title, using the format: `[_pt] PR: <description>`
- [x] Links are provided if this PR resolves an issue, or depends on another other PR
- [x] If submitting a PR to the `dev` branch (the default branch), you have a descriptive [Feature Branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) name using the format: `dev-<description-of-change>` (e.g.: `dev-revise-levee-masking`)
- [x] Changes are limited to a single goal (no scope creep)
- [x] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
- [ x Changes adhere to [PEP-8 Style Guidelines](https://peps.python.org/pep-0008/)
- [ ] (sort of) Any _change_ in functionality is tested (well.. except the "nws tools")
- [ ] (n/a) New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated ([CHANGELOG](/doc/CHANGELOG.md) and/or [README](/README.md))
- [x] [Reviewers requested](https://help.github.com/articles/requesting-a-pull-request-review/)

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+[project.urls]
+Source = "https://github.com/NOAA-OWP/ras2fim"
+Wiki = "https://github.com/NOAA-OWP/ras2fim/wiki"
+
+
+# - Tools -
+
+[tool.black]
+line_length = 110
+targer-version = ['py38']
```
Comment (robgpita-noaa) [inline pyproject.toml:38]: `target-version = ['py38']`?

Comment (RobHanna-NOAA) [reply pyproject.toml]: removed it as it has been tested and I assume it was ignored. I added it during some tests to upgrade python 3.8 but rolled it back.

**Review by JamesColl-NOAA — DISMISSED** [review]
image

This was quite extensive Rob, thanks for leading this format change.  The logging and extra output is noisy but results generated as expected and your PR notes were quite helpful.

**Review by JamesColl-NOAA — APPROVED** [review]
Math is deterministic :)

**Pull request merged**